Rubberduck.Core/AutoComplete/AutoCompleteService.cs

```diff
@@ -44,35 +44,71 @@ private void UpdateEnabledAutoCompletes(Configuration config)
             }
         }
 
-        private void VBENativeServices_CaretHidden(object sender, AutoCompleteEventArgs e)
+        private void HandleKeyDown(object sender, AutoCompleteEventArgs e)
         {
-            if (e.ContentHash == _contentHash)
+            if (e.Keys == Keys.Delete)
             {
                 return;
             }
 
-            var qualifiedSelection = e.CodePane.GetQualifiedSelection();
+            var module = e.CodeModule;
+            var qualifiedSelection = module.GetQualifiedSelection();
             var selection = qualifiedSelection.Value.Selection;
+            var currentContent = module.GetLines(selection);
 
             foreach (var autoComplete in _autoCompletes.Where(auto => auto.IsEnabled))
             {
-                if (autoComplete.Execute(e))
+                if (e.Keys == Keys.Back && selection.StartColumn > 1)
                 {
-                    _lastSelection = qualifiedSelection;
-                    _lastCode = e.NewCode;
-                    using (var module = e.CodePane.CodeModule)
+                    if (HandleBackspace(e, autoComplete))
                     {
-                        _contentHash = module.ContentHash();
+                        break;
                     }
+                }
+                else
+                {
+                    if (autoComplete.Execute(e))
+                    {
+                        break;
+                    }
+                }
+            }
+        }
 
-                    break;
+        private bool HandleBackspace(AutoCompleteEventArgs e, IAutoComplete autoComplete)
+        {
+            var isInlineAutoComplete = autoComplete.InputToken.Length == 1;
+            if (isInlineAutoComplete)
+            {
+                var code = e.CurrentLine;
+                // If caret LHS is the AC input token and RHS is the AC output token, we can remove both.
+                // Substring index is 0-based. Selection from code pane is 1-based.
+                // LHS should be at StartColumn - 2, RHS at StartColumn - 1.
+                var caretLHS = code.Substring(Math.Max(0, e.CurrentSelection.StartColumn - 2), 1);
+                var caretRHS = code.Length >= e.CurrentSelection.StartColumn
+                    ? code.Substring(e.CurrentSelection.StartColumn - 1, 1)
+                    : string.Empty;
+
+                if (caretLHS == autoComplete.InputToken && caretRHS == autoComplete.OutputToken
+                    /*&& !string.IsNullOrEmpty(code.Substring(e.CurrentSelection.StartColumn - 1))*/)
```

Comment:
remove

Reply:
To clarify will the whole procedure be removed or only that comment? I ask because there are references to carets all over the procedure. If they are no longer from carets it would be appropriate to rename them accordingly, no?